core/src/Kokkos_GraphNode.hpp

```diff
@@ -228,23 +229,44 @@ class GraphNodeRef {
 
   // TODO We should do better than a p-for (that uses registers, heavier).
   //      This should "just" launch the function on device with our driver.
-  template <typename Label, typename Functor,
+  template <typename Label, typename Policy, typename Functor,
             typename = std::enable_if_t<std::is_invocable_r_v<
                 void, const Kokkos::Impl::remove_cvref_t<Functor>>>>
-  auto then(Label&& label, const ExecutionSpace& exec,
+  auto then(Label&& label, const ExecutionSpace& exec, Policy&& policy,
             Functor&& functor) const {
+    static_assert(Kokkos::Impl::is_specialization_of_v<Policy, ThenPolicy>);
     using next_kernel_t =
-        Kokkos::Impl::GraphNodeThenImpl<ExecutionSpace,
+        Kokkos::Impl::GraphNodeThenImpl<ExecutionSpace, Policy,
                                         Kokkos::Impl::remove_cvref_t<Functor>>;
     return this->_then_kernel(next_kernel_t{std::forward<Label>(label), exec,
+                                            std::forward<Policy>(policy),
                                             std::forward<Functor>(functor)});
   }
 
-  template <typename Label, typename Functor,
+  // Overload for a label or policy given.
+  template <typename LabelOrPolicy, typename Functor,
             typename = std::enable_if_t<std::is_invocable_r_v<
                 void, const Kokkos::Impl::remove_cvref_t<Functor>>>>
-  auto then(Label&& label, Functor&& functor) const {
+  auto then(LabelOrPolicy&& label_or_policy, Functor&& functor) const {
+    if constexpr (Kokkos::Impl::is_specialization_of_v<LabelOrPolicy,
+                                                       ThenPolicy>) {
+      return this->then(std::string{}, ExecutionSpace{},
+                        std::forward<LabelOrPolicy>(label_or_policy),
+                        std::forward<Functor>(functor));
+    } else {
+      return this->then(std::forward<LabelOrPolicy>(label_or_policy),
+                        ExecutionSpace{}, ThenPolicy<>{},
```

Comment:
```suggestion
                        ExecutionSpace{}, ThenPolicy{},
```
(assuming we default the template parameter)

Reply:
This brings back the discussion that the exec space argument should be stored in the policy for consistency.

Reply:
* `ThenPolicy<>{}` -> `ThenPolicy{}` done :heavy_check_mark: 
* For why we don't want the `exec` in the policy, see #8191.